packages/backend/src/open_payments/wallet_address/middleware.ts

```diff
@@ -15,23 +15,24 @@ export async function getWalletAddressUrlFromRequestBody(
   ctx: SignedCollectionContext<CreateBody>,
   next: () => Promise<void>
 ) {
-  ctx.walletAddressUrl = ctx.request.body.walletAddress
+  ctx.walletAddressUrl = ctx.request.body.walletAddress.toLowerCase()
```

Comment:
Instead of doing this all over, we can just do `.toLowerCase` in  during creation (`walletAddressService.create`) and when we fetch it by url (`getWalletAddressByUrl`)

Reply:
If I correctly understood it would be better to discard the `packages/backend/src/open_payments/wallet_address/middleware.ts` and `packages/backend/src/graphql/resolvers/wallet_address.ts` changes and better lowercase inside `packages/backend/src/open_payments/wallet_address/service.ts`?

Reply:
@mkurapov maybe going a little bit further, but probably one very neat solution would be to use GraphQL directive for this kind of things?

Our input object would then look like this
```
input CreateWalletAddressInput {
  "Asset of the wallet address"
  assetId: String!
  "Wallet Address URL"
  url: String! @lower
  "Public name associated with the wallet address"
  publicName: String
  "Unique key to ensure duplicate or retried requests are processed only once. See [idempotence](https://en.wikipedia.org/wiki/Idempotence)"
  idempotencyKey: String
  "Additional properties associated with the [walletAddress]."
  additionalProperties: [AdditionalPropertyInput!]
}
```

Notice that `@lower`? With this kind of directive we would make sure that we always get lower case url in the request.

So the only question would be `getWalletAddressByUrl` right? If I am not mistaken this gets called only when we create receiver right or with other words, it is only getting called from GraphQL not Rest right? In that case, we can also add directive to that field.

Instead of doing this thing manually as close to the database as possible we would have lowercase url since the beginning and it would be the same as if user were inputing url in lower case.

But maybe this is too much for this ticket?

Reply:
@tadejgolobic 

That's quite a nice solution actually, and we probably have a few opportunities where we could leverage GQL directives more. 

For this case though, I think it makes sense to keep this in the service/domain layer because no matter the interface/API used to create wallet addresses, we always have this baked into the logic before we create the wallet address

Reply:
@tadejgolobic would this change not be retroactive? I mean the already created wallets will not be found. 

Assuming that you mean admin API instead of GraphQL. Yes, the `getWalletAddressByUrl` is geting called from the admin API and the rest (open payments>get method)